pkg/controller/templateupdaterequest/templateupdaterequest_controller.go

```diff
@@ -0,0 +1,320 @@
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileTemplateUpdateRequest) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	logger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	logger.Info("Reconciling TemplateUpdateRequest")
+
+	// Fetch the TemplateUpdateRequest tur
+	tur := &toolchainv1alpha1.TemplateUpdateRequest{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, tur)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// if is complete, then check when status was changed and delete it if the requested duration has passed
+	completeCond, found := condition.FindConditionByType(tur.Status.Conditions, toolchainv1alpha1.TemplateUpdateRequestComplete)
+	if found && completeCond.Status == corev1.ConditionTrue {
+		deleted, requeueAfter, err := r.checkTransitionTimeAndDelete(logger, tur, completeCond)
+		if deleted {
+			return reconcile.Result{}, nil
+		}
+		if err != nil {
+			return reconcile.Result{}, r.updateStatusConditions(logger, tur, toFailure(errs.Wrap(err, "unable to delete the resource")))
+		}
+		return reconcile.Result{
+			Requeue:      true,
+			RequeueAfter: requeueAfter,
+		}, nil
+	}
+
+	// lookup the MasterUserRecord with the same name as the TemplateUpdateRequest tur
+	mur := &toolchainv1alpha1.MasterUserRecord{}
+	err = r.client.Get(context.TODO(), request.NamespacedName, mur)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// MUR object not found, could have been deleted after reconcile request.
+			// Marking this TemplateUpdateRequest as failed
+			return reconcile.Result{}, r.updateStatusConditions(logger, tur, toFailure(err))
+		}
+		// Error reading the object - requeue the request.
+		logger.Error(err, "Unable to get the MasterUserRecord associated with the TemplateUpdateRequest")
+		return reconcile.Result{}, err
+	}
+	if len(tur.Status.SyncIndexes) == 0 {
+		// need to be "captured" before updating the MURs
+		syncIndexes := syncIndexes(tur.Spec.TierName, *mur)
+		// if the TemplateUpdateRequest was just created (ie, `Status.SyncIndexes` is empty), then we should update the associated MasterUserRecord
+		// and retain its current syncIndexex in the status
+		if err = r.updateTemplateRefs(logger, *tur, mur); err != nil {
+			logger.Error(err, "Unable to update the MasterUserRecord associated with the TemplateUpdateRequest")
+			return reconcile.Result{}, r.updateStatusConditions(logger, tur, toFailure(err))
```

Comment:
Do we want to set the status of the `TemplateUpdateRequest` to failed immeditately for the first error? What if we have another reason - eg. `UpdateErrored` - that is set for the first failure, then if the status already contains the `UpdateErrored` reason, then it would set it to: `UnableToUpdate` which is understood as a failed status. In the future, we may add a number of errors, but for now we could add at least one attempt.

Reply:
I understand the idea and I agree with you that the controller should be permissive and insist/retry a few times, but can we address this in a separate PR?

Reply:
this will be addressed in https://issues.redhat.com/browse/CRT-708